Site Domains: Modernize UI 2/2 - Integrate status and navigation to domain details

staskus opened the pull request:
Fixes #https://github.com/wordpress-mobile/WordPress-iOS/issues/22262

Continuation of https://github.com/wordpress-mobile/WordPress-iOS/pull/22294

## Description

Use `all-domains` endpoint instead of `sites/%d/domains` to load information needed to display status and open domain details.

## Solution

1. Use `DomainsService` to fetch domains and filter to show domains for a specific site
2. Cache fetched domains to preload them when opening All Domains
3. Add loading and error views
4. Add NavigationLink to open domain details web view

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/8c838b67-7e70-4049-b0e9-19343aadc0b5

## To test:

1. Check `Menu -> Domains` UI:
- With an account that has a domain (domains)
- With an account that only has a free domain
2. Confirm Status indicator appears
3. Confirm Status indicator matches the one on All Domains screen
4. Confirm tapping on paid domains opens Domain Management

## Regression Notes
1. Potential unintended areas of impact

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

staskus committed:
Load domains using all-domains endpoint

Combine domainsList and all-domains to show a more comprehensive site domains list that is equivalent to All Domains view

staskus committed:
Create DomainsStateView for empty and error state

staskus committed:
Reuse error handling between AllDomainsListViewModel and SiteDomainsViewModel

staskus committed:
Integrate DomainsStateView into SiteDomainsView

staskus committed:
Add ProgressView() to SiteDomainsView

staskus committed:
Pass loaded domains from SiteDomains to AllDomains

Avoid additional loading of domains when navigating from SiteDomains to AllDomains

staskus committed:
Open DomainDetails after selecting purchased domain

staskus committed:
Added SiteDomainsViewModelTests

staskus committed:
Update RELEASE-NOTES.txt

irfano commented:
The status label and navigating to domain detail worked as expected. 🟢 

#### A minor issue
The domain detail screen should have a title when navigating from Site Domains.

|from All Domains|from Site Domains|
|-|-|
|![Simulator Screenshot - iPhone 15 Pro Max - 2024-01-03 at 14 18 43](https://github.com/wordpress-mobile/WordPress-iOS/assets/2471769/6be1804d-d1d0-4d1b-9b10-ed08b01747f0)|![Simulator Screenshot - iPhone 15 Pro Max - 2024-01-03 at 14 18 03](https://github.com/wordpress-mobile/WordPress-iOS/assets/2471769/a29417eb-9a50-4f86-96bd-61bbc4eb7893)|

#### Touch area for the "Add a domain" button
This looks like unrelated to this PR. If so, I can open an issue for it.

https://github.com/wordpress-mobile/WordPress-iOS/assets/2471769/5e973138-004f-4fd3-a42a-a2036ad14b5f

#### Platform difference
When there is both a custom domain and credits, iOS shows a "Claim your free domain" button, while Android shows a card. I think Android's way is better. We can open an issue to match platforms later.

|iOS|Android|
|-|-|
|<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/2471769/001a232b-5d5c-477c-8f32-18d2a474b63e" width=300>|<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/2471769/5611e7b5-f071-4484-a959-9bd7ce273cef" width=300>|

#### Other domains title
The title is changed on the design to "OTHER DOMAINS FOR STOP MAKING SENSE". 

#### Color issue on Webview in dark mode
This is already existed on trunk, opened issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/22324

staskus commented:
@irfano, thanks for testing it!

> The domain detail screen should have a title when navigating from Site Domains.

I'll fix it.

> Touch area for the "Add a domain" button

Good catch. I will see what can be done with it. I would also expect the touch area to take the area of the whole row.

> The title is changed on the design to "OTHER DOMAINS FOR STOP MAKING SENSE".

I'll fix it.

> When there is both a custom domain and credits, iOS shows a "Claim your free domain" button, while Android shows a card. I think Android's way is better. We can open an issue to match platforms later

Good catch 👍 . It should be an easy fix.

staskus committed:
Set title to Domain Management view

staskus committed:
Use site name for other domains section title

guarani reviewed (commented):
The domain status label sometimes shows Activating when WP.com shows Active. `/all-domains?resolve_status=true&no_wpcom=true` returns "Activating" for that domain, which is unexpected since the domain wasn't registered recently. My best guess is that "Active" is the right status. I don't see this mismatch on all my domains, do you see it on any of yours?

Calypso has frontend logic for displaying the status here: https://github.com/Automattic/wp-calypso/blob/7b4243596a876949629fc5bfa2c7a7e8566478b8/client/lib/domains/resolve-domain-status.tsx#L56C17-L56C36

I'd need to spend more time on this to figure out how to handle this. We could address this issue separately to this PR since I see the same mismatch on the All Domains screen.

---

In the video attached to the PR, the domain changes from povilasstaskus.wordpress.com to povilasstaskus.wpcomstaging.com when you view the domain details web view (it then changes back after you buy the new domain). I didn't see this on my domains. Can you reproduce or could this be an issue that was resolved in a later commit?

---

If you want to address this separately, I can continue reviewing this PR tomorrow as I don't have time to finish it today.

staskus commented:
Thanks for the review, @guarani !

> The domain status label sometimes shows Activating when WP.com shows Active. `/all-domains?resolve_status=true&no_wpcom=true` returns "Activating" for that domain, which is unexpected since the domain wasn't registered recently. My best guess is that "Active" is the right status. I don't see this mismatch on all my domains, do you see it on any of yours?
> 
> Calypso has frontend logic for displaying the status here: https://github.com/Automattic/wp-calypso/blob/7b4243596a876949629fc5bfa2c7a7e8566478b8/client/lib/domains/resolve-domain-status.tsx#L56C17-L56C36

Yes, Calypso looks to have domain status-resolving logic within a client, which is strange to me. The backend through this endpoint resolves domain status as well fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Syvo%2Sqbznvaf%2Sqbznva%2Qfgnghf%2Qerfbyire%2Sqbznva%2Qfgnghf%2Qerfbyire.cuc%3Se%3Qsn1n97n3%23341-og . It could be activating if "pending_registration" flag exists. fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Syvo%2Sqbznvaf%2Sjcpbz%2Qqbznva.cuc%3Se%3Q59p4n8r9%231793-og. 

I don't know why this mismatch happens as well, I didn't dig deep into domain management logic itself, it's complicated. 

> We could address this issue separately to this PR since I see the same mismatch on the All Domains screen.

Yes, I think it would be best to address it separately. It may not be addressable from the app. Maybe there're some differences between Calypso frontend and backend and how the status is resolved.

> Can you reproduce or could this be an issue that was resolved in a later commit?

Oh yes, I see it now. I can reproduce it 🤔 `.freeSiteAddress` is different at different times it's checked. 

It happens because `domains` Set contains multiple `wpCom` domains, one `.wordPress.com`, and one `.wpcomstaging.com`. We take `.first` as freeDomain and since `Set` doesn't guarantee an order, we get a different one each time. 

This issue happens on a previous Site Domains view as well:

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/81022ebd-569e-491d-92e7-ab6ad223f388

> If you want to address this separately, I can continue reviewing this PR tomorrow as I don't have time to finish it today.

I think domain statuses should be investigated separately, I'm not sure anything could be done in the scope of this PR.

staskus committed:
Show new domain card section also when blog has domain credit

staskus committed:
Make add domain button touch area full row

staskus committed:
Load free domains from allDomains endpoint

AllDomains endpoint result allows to distinguish between a simple and staging free domain

staskus commented:
Made fixes:

- [x] The domain detail screen should have a title when navigating from Site Domains
- [x] Touch area for the "Add a domain" button
- [x] The title is changed on the design to "OTHER DOMAINS FOR STOP MAKING SENSE".
- [x] When there is both a custom domain and credits, show a "Claim your free domain" card
- [x] .freeSiteAddress is different at different times it's checked

guarani commented on WordPress/Classes/ViewRelated/Domains/View Models/SiteDomainsViewModel.swift:
```diff
@@ -115,7 +159,54 @@ private extension SiteDomainsViewModel {
+        struct Button {
+            let title: String
+            let action: () -> Void
+        }
```
nitpick: It looks like `DomainsStateViewModel` replaced this and this struct can be removed.

guarani commented on WordPress/Classes/ViewRelated/Domains/Views/DomainsStateView.swift:
nitpick: This new domain state view has a very similar layout to the existing `AllDomainsListEmptyView`. Would it make sense to consolidate? I think `DomainStateView` would be fine to reuse.

| `AllDomainsListEmptyView` | `DomainStateView` |
|-|-|
| <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/ee66beac-4fbb-43fb-a69f-20f62cebac2c" width="350"> | <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/d2d3e3ea-5eaa-4588-b038-3f841d84e23a" width="350"> |

staskus replied:
I agree with you, good point. When I developed it, I admit I tried to reuse the on from All Domains view and it appeared with broken spacing within the Swift UI layout, no matter how much I tried to constrain it within HStacks and VStacks. I got frustrated and created a new SwiftUI view. But I can try the opposite way, reuse SwiftUI view within UIKit. Maybe it will appear correctly 👍

staskus replied:
Fixed!

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/fe340caa-b7eb-4e9a-a045-766155c9b678

guarani replied:
Nice!

guarani commented on WordPress/WordPressTest/Domains/SiteDomainsViewModelTests.swift:
👏 It's great this level of test coverage!

guarani reviewed (approved):
I re-tested and didn't find any issues and the code comments aren't blockers. Thanks for the patience on this one, @staskus 🙇

guarani commented on WordPress/Classes/ViewRelated/Domains/View Models/SiteDomainsViewModel.swift:
```diff
@@ -1,112 +1,156 @@
-    private static func buildSections(from blog: Blog) -> [Section] {
-        return Self.buildFreeDomainSections(from: blog) + Self.buildDomainsSections(from: blog)
+    private static func buildSections(from blog: Blog, domains: [DomainsService.AllDomainsListItem]) -> [Section] {
+        var wpcomDomains: [DomainsService.AllDomainsListItem] = []
+        var otherDomains: [DomainsService.AllDomainsListItem] = []
+
+        for domain in domains {
+            if domain.wpcomDomain {
+                wpcomDomains.append(domain)
+            } else {
+                otherDomains.append(domain)
+            }
+        }
```
nitpick: This could be easier to read:

```suggestion
        let wpcomDomains = domains.filter { $0.wpcomDomain }
	let otherDomains = domains.filter { !$0.wpcomDomain }
```

guarani commented on WordPress/Classes/ViewRelated/Domains/View Models/SiteDomainsViewModel.swift:
```diff
@@ -1,112 +1,156 @@
-        let primaryDomain = blog.domainsList.first(where: { $0.domain.isPrimaryDomain })
-        let otherDomains = blog.domainsList.filter { !$0.domain.isPrimaryDomain }
+        let primaryDomainName = blog.domainsList.first(where: { $0.domain.isPrimaryDomain })?.domain.domainName
+        var primaryDomain: DomainsService.AllDomainsListItem?
+        var otherDomains: [DomainsService.AllDomainsListItem] = []
+
+        for domain in domains {
+            if domain.blogId == blog.dotComID?.intValue {
+                if primaryDomainName == domain.domain {
+                    primaryDomain = domain
+                } else {
+                    otherDomains.append(domain)
+                }
+            }
+        }
```
nitpick: This alternative could also work:
```suggestion
        let blogDomains = domains.filter({ $0.blogId == blog.dotComID?.intValue })
        let primaryDomain = blogDomains.first(where: { primaryDomainName == $0.domain })
        let otherDomains = blogDomains.filter({ primaryDomainName != $0.domain })
```

staskus replied:
I did all this to have fewer iterations but that's probably unneeded optimization 😀. 2N or 3N is still N. There're no nested loops here. I will use your proposal 👍

staskus committed:
Remove unused MessagateStateViewModel

staskus committed:
Update SiteDomainsViewModel.swift

staskus committed:
Use DomainsStateView within AllDomainsView
